Cypress Setup

revanth0212 opened the pull request:
## Description

Initial cypress setup.

## Related Issue
Closes PWA-1569

### Verification Stakeholders
@dpatil-magento 

### Specification
<!-- Changes to `upward-spec` and/or `upward-js` packages must be reviewed -->
<!-- by `UPWARD-PHP` maintainers to ensure continued compatibility -->

### Verification Steps
1. Checkout the branch. Navigate to `pwa-automation-tests`.
2. Run `yarn install` if its your first time.
3. Run `yarn test` to open up the cypress UI. You should see 2 dummy tests but they won't run. They are just provided as examples.

## Screenshots / Screen Captures (if appropriate)

![image](https://user-images.githubusercontent.com/35203638/112367424-e2f3fd00-8ca7-11eb-9d1a-7e671450762e.png)

## Checklist
<!--- Go over all the following points, and make sure you've done anything necessary -->
* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.

dpatil-magento reviewed (changes requested):
Looking for examples on how to manage 

- css element declaration in one place and call them as needed in tests.
- where to add common actions like login, place order and call them in tests as needed.
- where to mock data

revanth0212 commented:
Addressed the PR comments. Created a new folder structure. But unfortunately, the sample tests I left in the PR are dummy ones. Anyways we will add new tests once this is merged.

dpatil-magento commented:
@revanth0212 Both sample tests are failing, can you please check?

supernova-at commented on venia-integration-tests/README.md:
```diff
@@ -0,0 +1,7 @@
+# Running Venia Functional Tests
+
+## Setup
+1. Clone the repo and change directory to pwa-studio/venia-e2e-tests
```
```suggestion
1. Clone the repo and change directory to pwa-studio/venia-integration-tests
```

supernova-at commented on venia-integration-tests/README.md:
```diff
@@ -0,0 +1,7 @@
+## Setup
+1. Clone the repo and change directory to pwa-studio/venia-e2e-tests
+2. Run `yarn install`
+3. Run `yarn cypress open`
```
```suggestion
3. Run `yarn test`
```

supernova-at commented on venia-integration-tests/cypress.config.json:
```diff
@@ -0,0 +1,12 @@
+    "integrationFolder": "src/tests",
+    "pageLoadTimeout": 30000,
+    "requestTimeout": 30000,
+    "testFiles": "**/*.spec.js",
```
May want to consider restricting this to the `src/tests` folder, but probably fine

revanth0212 replied:
Yes, it checks for *.spec.js in the `integrationFolder`. So basically we are checking for those files in the `src/tests` folder.

supernova-at commented on venia-integration-tests/cypress.config.json:
```diff
@@ -0,0 +1,12 @@
+    "pageLoadTimeout": 30000,
+    "requestTimeout": 30000,
+    "testFiles": "**/*.spec.js",
+    "viewportHeight": 900,
+    "viewportWidth": 1440
```
Can we have different configurations if we wanted to run mobile- or desktop-specific tests?

revanth0212 replied:
The problem is we don't have a single definition of mobile so I left it here. Depending on the tests we can set the viewport dimensions per test instead of a global value.

supernova-at commented on venia-integration-tests/cypress/plugins/index.js:
```diff
@@ -0,0 +1,21 @@
+// This example plugins/index.js can be used to load plugins
+//
+// You can change the location of this file or turn off loading
+// the plugins file with the 'pluginsFile' configuration option.
```
Do we need this file at this time?

supernova-at replied:
Same with `cypress/support` files

revanth0212 replied:
Yeah you are right, not sure how I missed deleting this folder. Thanks Andy.

supernova-at commented on venia-integration-tests/package.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "pwa-automation-tests",
```
Consistency - some places it's `pwa-automation-tests`, `venia-e2e-tests`, or `venia-integration-tests`.

supernova-at commented on venia-integration-tests/package.json:
```diff
@@ -0,0 +1,17 @@
+        "test:firefox": "cypress open --browser firefox --config-file cypress.config.json",
+        "test:headless": "cypress open --browser chrome --headless --config-file cypress.config.json",
+        "test:run": "cypress run --browser chrome --config-file cypress.config.json",
+        "test:debug": "node --inspect-brk ./node_modules/cypress/bin/cypress open --browser chrome --config-file cypress.config.json"
```
Nice 👍

supernova-at commented on venia-integration-tests/src/fields/accountMenu/index.js:
```diff
@@ -0,0 +1,3 @@
+export const accountMenuItemsField = 'div[data-id="account-menu-items"]';
```
Consider using a slightly more specific name for this data attribute to avoid collisions.

revanth0212 replied:
I added this as a sample. We don't have `data-ids` in our code base yet. These are just dummy ones. When each dev decides to add new tests they can update these files accordingly.

supernova-at reviewed (changes requested):
Awesome! Left some considerations for you.

Also, I think I am missing a commit? I don't see the new `data-id` attribute added to any components, so the tests are failing:

![Screen Shot 2021-04-01 at 1 27 17 PM](https://user-images.githubusercontent.com/13182778/113339169-b5363600-92ef-11eb-96f8-653917a74128.png)

revanth0212 commented:
> Also, I think I am missing a commit? I don't see the new `data-id` attribute added to any components, so the tests are failing:

You are right @supernova-at there are no `data-id`s yet. This is a sample folder structure with sample tests that don't run 😅  like mentioned in this comment: https://github.com/magento/pwa-studio/pull/3082#issuecomment-809791997

Should have added this in the description. My bad.
